Add a disassembler

zenhack opened the pull request:
I wrote this while debugging the multi-assets work. There are some
things left to do:

- I need to properly document its interface, mainly the exact format of
  the return value.
- There is some code to try to detect when a PUSH is pushing a value
  corespodning to the address of a JUMPDEST, and record this fact, but
  It doesn't appear to be working; I should either figure out why or
  just punt and drop this feature.

...but I wanted to get this recorded somewhere before calling it a day.

fare commented:
Looks generally fine to me. See @AlexKnauth 's suggestion for a better handling of PUSH. String manipulation is better avoided.

zenhack commented:
Alright, I've:

- Added a test
- Incorporated your suggestion, pulling in code from @AlexKnauth implementation on discord
- Dropped the WIP on detecting label references; we can add that later if we want.
- Added documentation.

Marking this as not-a-draft.

fare commented on assembly.ss:
```diff
@@ -100,7 +102,47 @@
+           (let ()
+             (def rest (cdr data))
+             (def arg-bytes (take rest push-amt))
+             (def arg-hex (string-append "#x" (hex-encode (list->u8vector arg-bytes))))
```
Since we're in Ethereum land, I'd use "0x" instead of "#x" as the prefix.

fare commented on assembly.ss:
```diff
@@ -100,7 +102,47 @@
+;;   both forms readily available is sometimes helpful when debugging.
+(def (disassemble bytes)
+  (def labels (make-hash-table))
+  (let loop ((data (u8vector->list bytes)))
```
I would be more efficient to loop with an index.

Considering that a program is 24KB max, that performance isn't necessary to squeeze right now — however, that deserves at least a TODO comment in the code.

fare reviewed (approved):
LGTM modulo minor nits

fare commented:
Oh, and the tests should probably involve such degenerate cases as pushing extra zeroes in front, as in `PUSH3 42 "0x00002a"`

zenhack commented:
Done.
